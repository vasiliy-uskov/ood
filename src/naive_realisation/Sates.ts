export enum State {
	SoldOut,
	NoQuarter,
	HasQuarter,
}


export function stateToString(state: State) {
	switch (state) {
		case State.HasQuarter:
			return "waiting for turn of crank";
		case State.SoldOut:
			return "sold out";
		case State.NoQuarter:
			return "waiting for quarter";
	}
}