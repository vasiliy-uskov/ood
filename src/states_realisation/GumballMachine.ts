import {IState} from "./IState";
import {IGumballMachine} from "./IGumballMachine";
import {SoldOutState} from "./SoldOutState";
import {NoQuarterState} from "./NoQuarterState";
import {HasQuarterState} from "./HasQuarterState";
import {ILogger} from "../ILogger";

export class GumballMachine {
	private readonly _logger: ILogger;
	private readonly _soldOutState: SoldOutState;
	private readonly _noQuarterState: NoQuarterState;
	private readonly _hasQuarterState: HasQuarterState;
	private _state: IState;
	private _gumballsCount: number;
	private _quartersCount = 0;

	constructor(gumballsCount: number, logger: ILogger) {
		this._logger = logger;
		const interfaceImpl = this._getGumballMachineInterface();
		this._soldOutState = new SoldOutState(interfaceImpl, logger);
		this._noQuarterState = new NoQuarterState(interfaceImpl, logger);
		this._hasQuarterState = new HasQuarterState(interfaceImpl, logger);
		this.refill(gumballsCount)
	}

	ejectQuarter() {
		this._state.ejectQuarter();
	}

	insertQuarter() {
		this._state.insertQuarter();
	}

	turnCrank() {
		this._state.turnCrank();
	}

	refill(count: number) {
		this._gumballsCount = count;
		if (!this._gumballsCount) {
			this._setSoldOutState();
		}
		else if (this._quartersCount) {
			this._setHasQuarterState();
		}
		else {
			this._setNoQuarterState();
		}
	}

	toString(): string {
		return `
Mighty Gumball, Inc.
JS-enabled Standing Gumball Model #2019
Inventory: ${this._gumballsCount} gumball${this._gumballsCount != 1 ? "s" : ""}
Machine is ${this._state.toString()}
`;
	}

	private _getGumballMachineInterface(): IGumballMachine {
		return {
			getQuartersCount: this._getQuartersCount.bind(this),
			insertQuarter: this._insertQuarter.bind(this),
			ejectQuarters: this._ejectQuarters.bind(this),
			getBallCount: this._getBallCount.bind(this),
			releaseBall: this._releaseBall.bind(this),
			setSoldOutState: this._setSoldOutState.bind(this),
			setNoQuarterState: this._setNoQuarterState.bind(this),
			setHasQuarterState: this._setHasQuarterState.bind(this),
		}
	}

	private _getQuartersCount(): number {
		return this._quartersCount;
	}

	private _insertQuarter(): void {
		++this._quartersCount;
	}

	private _ejectQuarters(): void {
		this._quartersCount = 0;
	}

	private _getBallCount(): number {
		return this._gumballsCount;
	}

	private _releaseBall(): void {
		--this._gumballsCount;
		--this._quartersCount;
	}

	private _setSoldOutState(): void {
		this._state = this._soldOutState;
	}

	private _setNoQuarterState(): void {
		this._state = this._noQuarterState;
	}

	private _setHasQuarterState(): void {
		this._state = this._hasQuarterState;
	}
}