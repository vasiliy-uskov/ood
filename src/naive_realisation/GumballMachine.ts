import {State, stateToString} from "./Sates";
import {ILogger} from "../ILogger";

export class GumballMachine {
	private _state: State;
	private _count: number;
	private _logger: ILogger;

	constructor(count: number, logger: ILogger) {
		this._state = count > 0 ? State.NoQuarter : State.SoldOut;
		this._count = count;
		this._logger = logger;
	}

	insertQuarter(): void {
		switch (this._state) {
			case State.SoldOut:
				this._logger.log("You can't insert a quarter, the machine is sold out");
				break;
			case State.NoQuarter:
				this._logger.log("You inserted a quarter");
				this._state = State.HasQuarter;
				break;
			case State.HasQuarter:
				this._logger.log("You can't insert another quarter");
				break;
		}
	}

	ejectQuarter(): void {
		switch (this._state) {
			case State.HasQuarter:
				this._logger.log("Quarter returned");
				this._state = State.NoQuarter;
				break;
			case State.NoQuarter:
				this._logger.log("You haven't inserted a quarter");
				break;
			case State.SoldOut:
				this._logger.log("You can't eject, you haven't inserted a quarter yet");
				break;
		}
	}

	turnCrank(): void {
		switch (this._state) {
			case State.SoldOut:
				this._logger.log("You turned but there's no gumballs");
				break;
			case State.NoQuarter:
				this._logger.log("You turned but there's no quarter");
				break;
			case State.HasQuarter:
				this._logger.log("You turned...");
				this._logger.log("A gumball comes rolling out the slot...");
				--this._count;
				if (this._count == 0) {
					this._logger.log("Oops, out of gumballs");
					this._state = State.SoldOut;
				}
				else {
					this._state = State.NoQuarter;
				}
				break;
		}
	}

	refill(numBalls: number): void {
		this._count = numBalls;
		this._state = numBalls > 0 ? State.NoQuarter : State.SoldOut;
	}

	toString(): string {
		return `
Mighty Gumball, Inc.
JS-enabled Standing Gumball Model #2019
Inventory: ${this._count} gumball${this._count != 1 ? "s" : ""}
Machine is ${stateToString(this._state)}
`;
	}
}