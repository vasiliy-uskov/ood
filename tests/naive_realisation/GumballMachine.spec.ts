import {MockLogger} from "../MockLogger";
import {GumballMachine} from "../../src/naive_realisation/GumballMachine";

function useCase(message: string, startGumballsCount: number, useCaseFn: (machine: GumballMachine) => void) {
	it(message, () => {
		const logger = new MockLogger();
		const machine = new GumballMachine(startGumballsCount, logger);
		useCaseFn(machine);
		logger.matchResult();
		expect(machine.toString()).toMatchSnapshot();
	});
}

useCase('Если вставить четвертак и провернуть ручку, машина отдаёт жвачку', 1, machine => {
	machine.insertQuarter();
	machine.turnCrank();
});

useCase('После первого шарика жвачки можно получить следующий, если в автомате есть жвачка', 2, machine => {
	machine.insertQuarter();
	machine.turnCrank();
	machine.insertQuarter();
	machine.turnCrank();
});

useCase('Можно забрать четвертак, после того как он был вставлен', 1, machine => {
	machine.insertQuarter();
	machine.ejectQuarter();
});

useCase('Можно заполнить автомат', 0, machine => {
	machine.refill(1);
});

useCase('Можно достать всю жвачку из автомата', 2, machine => {
	machine.refill(0);
});

useCase('Если ручка была провёрнута четвертак изъять нельзя', 1, machine => {
	machine.insertQuarter();
	machine.turnCrank();
	machine.ejectQuarter();
});

useCase('Нельзя вставить два четвертака за раз', 2, machine => {
	machine.insertQuarter();
	machine.insertQuarter();
});

useCase('Нельзя достать четвертак, если он не был всавлен', 2, machine => {
	machine.ejectQuarter();
});


useCase('Нельзя провернуть ручку, если четвертак не был вставлен', 1, machine => {
	machine.turnCrank();
});

useCase('Нельзя получить жвачку если она закочилась', 0, machine => {
	machine.insertQuarter();
	machine.turnCrank();
});