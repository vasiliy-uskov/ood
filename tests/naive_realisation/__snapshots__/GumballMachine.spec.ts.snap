// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Если вставить четвертак и провернуть ручку, машина отдаёт жвачку 1`] = `
"You inserted a quarter

You turned...

A gumball comes rolling out the slot

Oops, out of gumballs

"
`;

exports[`Если вставить четвертак и провернуть ручку, машина отдаёт жвачку 2`] = `
"
Mighty Gumball, Inc.
JS-enabled Standing Gumball Model #2016
Inventory: 0 gumballs
Machine is sold out
"
`;

exports[`Если ручка была провёрнута четвертак изъять нельзя 1`] = `
"You inserted a quarter

You turned...

A gumball comes rolling out the slot

Oops, out of gumballs

You can't eject, you haven't inserted a quarter yet

"
`;

exports[`Если ручка была провёрнута четвертак изъять нельзя 2`] = `
"
Mighty Gumball, Inc.
JS-enabled Standing Gumball Model #2016
Inventory: 0 gumballs
Machine is sold out
"
`;

exports[`Можно достать всю жвачку из автомата 1`] = `""`;

exports[`Можно достать всю жвачку из автомата 2`] = `
"
Mighty Gumball, Inc.
JS-enabled Standing Gumball Model #2016
Inventory: 0 gumballs
Machine is sold out
"
`;

exports[`Можно забрать четвертак, после того как он был вставлен 1`] = `
"You inserted a quarter

Quarter returned

"
`;

exports[`Можно забрать четвертак, после того как он был вставлен 2`] = `
"
Mighty Gumball, Inc.
JS-enabled Standing Gumball Model #2016
Inventory: 1 gumball
Machine is waiting for quarter
"
`;

exports[`Можно заполнить автомат 1`] = `""`;

exports[`Можно заполнить автомат 2`] = `
"
Mighty Gumball, Inc.
JS-enabled Standing Gumball Model #2016
Inventory: 1 gumball
Machine is waiting for quarter
"
`;

exports[`Нельзя вставить два четвертака за раз 1`] = `
"You inserted a quarter

You can't insert another quarter

"
`;

exports[`Нельзя вставить два четвертака за раз 2`] = `
"
Mighty Gumball, Inc.
JS-enabled Standing Gumball Model #2016
Inventory: 2 gumballs
Machine is waiting for turn of crank
"
`;

exports[`Нельзя достать четвертак, если он не был всавлен 1`] = `
"You haven't inserted a quarter
"
`;

exports[`Нельзя достать четвертак, если он не был всавлен 2`] = `
"
Mighty Gumball, Inc.
JS-enabled Standing Gumball Model #2016
Inventory: 2 gumballs
Machine is waiting for quarter
"
`;

exports[`Нельзя получить жвачку если она закочилась 1`] = `
"You can't insert a quarter, the machine is sold out

You turned but there's no gumballs

"
`;

exports[`Нельзя получить жвачку если она закочилась 2`] = `
"
Mighty Gumball, Inc.
JS-enabled Standing Gumball Model #2016
Inventory: 0 gumballs
Machine is sold out
"
`;

exports[`Нельзя провернуть ручку, если четвертак не был вставлен 1`] = `
"You turned but there's no quarter

"
`;

exports[`Нельзя провернуть ручку, если четвертак не был вставлен 2`] = `
"
Mighty Gumball, Inc.
JS-enabled Standing Gumball Model #2016
Inventory: 1 gumball
Machine is waiting for quarter
"
`;

exports[`После первого шарика жвачки можно получить следующий, если в автомате есть жвачка 1`] = `
"You inserted a quarter

You turned...

A gumball comes rolling out the slot

You inserted a quarter

You turned...

A gumball comes rolling out the slot

Oops, out of gumballs

"
`;

exports[`После первого шарика жвачки можно получить следующий, если в автомате есть жвачка 2`] = `
"
Mighty Gumball, Inc.
JS-enabled Standing Gumball Model #2016
Inventory: 0 gumballs
Machine is sold out
"
`;
